File: src/temporal/activities/submissions.py
"""Submission-related activities for Temporal bot."""

from datetime import datetime, timezone
from types import SimpleNamespace

from temporalio import activity

from bot.config import MONTHLY_POST_FLAIR_ID, SUBREDDIT_NAME
from bot.models import ActiveSubmissions, CreateMonthlyPostInput, SubmissionInput
from bot.reddit import get_bot_user, get_reddit_client, get_subreddit
from .helpers import TemplateManager


@activity.defn
def fetch_active_submission_ids() -> ActiveSubmissions:
    """Discover current and previous submission IDs from Reddit.

    Used once at polling workflow startup to bootstrap state.
    Stickied submission = current; most recent non-stickied, non-locked = previous.
    """
    reddit = get_reddit_client()
    bot_user = get_bot_user(reddit)

    current_id: str | None = None
    previous_id: str | None = None

    for submission in bot_user.submissions.new(limit=5):
        if submission.stickied and current_id is None:
            current_id = submission.id
        elif not submission.stickied and not submission.locked and previous_id is None:
            previous_id = submission.id

        if current_id is not None and previous_id is not None:
            break

    activity.logger.info(
        "Discovered submissions: current=%s, previous=%s", current_id, previous_id
    )
    return ActiveSubmissions(
        current_submission_id=current_id,
        previous_submission_id=previous_id,
    )


@activity.defn
def sticky_submission(input: SubmissionInput) -> bool:
    """Sticky a specific submission. Idempotent."""
    reddit = get_reddit_client()
    submission = reddit.submission(id=input.submission_id)

    if not submission.stickied:
        submission.mod.sticky(bottom=False)
        activity.logger.info(
            "Stickied: https://reddit.com%s", submission.permalink
        )

    return True


@activity.defn
def unsticky_submission(input: SubmissionInput) -> bool:
    """Unsticky a specific submission. Idempotent."""
    reddit = get_reddit_client()
    submission = reddit.submission(id=input.submission_id)

    if submission.stickied:
        submission.mod.sticky(state=False)
        activity.logger.info(
            "Unstickied: https://reddit.com%s", submission.permalink
        )

    return True


@activity.defn
def lock_submission(input: SubmissionInput) -> bool:
    """Lock a specific submission. Idempotent."""
    reddit = get_reddit_client()
    submission = reddit.submission(id=input.submission_id)

    if not submission.locked:
        submission.mod.lock()
        activity.logger.info(
            "Locked: https://reddit.com%s", submission.permalink
        )

    return True


@activity.defn
def create_monthly_post(input: CreateMonthlyPostInput) -> str:
    """Create a new monthly confirmation thread.

    This activity is idempotent: if a post for the current month already exists,
    returns that submission ID without creating a duplicate.

    No longer stickies the post — that is handled by a separate activity.

    Returns the submission ID.
    """
    reddit = get_reddit_client()
    subreddit = get_subreddit(reddit)
    bot_user = get_bot_user(reddit)

    now = datetime.now(timezone.utc)

    # Fetch previous submission metadata for idempotency check and template context.
    previous_submission: dict | None = None
    if input.previous_submission_id:
        sub = reddit.submission(id=input.previous_submission_id)
        previous_submission = {
            "id": sub.id,
            "title": sub.title,
            "permalink": sub.permalink,
            "created_utc": sub.created_utc,
        }
    else:
        try:
            last_post = next(bot_user.submissions.new(limit=1))
            previous_submission = {
                "id": last_post.id,
                "title": last_post.title,
                "permalink": last_post.permalink,
                "created_utc": last_post.created_utc,
            }
        except StopIteration:
            pass

    # Idempotency check: see if we already created a post this month
    if previous_submission is not None:
        post_date = datetime.fromtimestamp(previous_submission["created_utc"], tz=timezone.utc)
        if post_date.year == now.year and post_date.month == now.month:
            activity.logger.info(
                "Monthly post already exists for this month (%s), returning existing ID (idempotent)",
                previous_submission["id"],
            )
            return previous_submission["id"]

    if previous_submission:
        template_submission = SimpleNamespace(
            title=previous_submission["title"],
            permalink=previous_submission["permalink"],
        )
    else:
        template_submission = SimpleNamespace(
            title="Previous monthly thread",
            permalink=f"https://www.reddit.com/r/{SUBREDDIT_NAME}/",
        )

    # Load templates
    post_template = TemplateManager.load("monthly_post", subreddit)
    title_template = TemplateManager.load("monthly_post_title", subreddit)

    activity.logger.info("Creating monthly post for r/%s", SUBREDDIT_NAME)

    # Create new post
    new_submission = subreddit.submit(
        title=now.strftime(title_template),
        selftext=post_template.format(
            bot_name=bot_user.name,
            subreddit_name=SUBREDDIT_NAME,
            previous_month_submission=template_submission,
            now=now,
        ),
        flair_id=MONTHLY_POST_FLAIR_ID,
        send_replies=False,
    )

    new_submission.mod.suggested_sort(sort="new")

    activity.logger.info(
        "Created monthly post: https://reddit.com%s", new_submission.permalink
    )
    return new_submission.id

File: internal/activities/submissions.go
package activities

import (
	"context"
	"fmt"
	"net/http"
	"net/url"
	"os"
	"time"

	"go.temporal.io/sdk/activity"

	"github.com/mikeacjones/reddit-trade-confirmation-bot/internal/models"
	"github.com/mikeacjones/reddit-trade-confirmation-bot/internal/templates"
)

// FetchActiveSubmissionIDs discovers current and previous submission IDs from Reddit.
func (a *Activities) FetchActiveSubmissionIDs(ctx context.Context) (models.ActiveSubmissions, error) {
	subs, err := a.Reddit.BotSubmissions(5)
	if err != nil {
		return models.ActiveSubmissions{}, err
	}
	var current, previous *string
	for _, s := range subs {
		id := s.ID
		if s.Stickied && current == nil {
			current = &id
		} else if !s.Stickied && !s.Locked && previous == nil {
			previous = &id
		}
		if current != nil && previous != nil {
			break
		}
	}
	activity.GetLogger(ctx).Info("Discovered submissions",
		"current", ptrStr(current), "previous", ptrStr(previous))
	return models.ActiveSubmissions{
		CurrentSubmissionID:  current,
		PreviousSubmissionID: previous,
	}, nil
}

// StickySubmission stickies a submission (idempotent).
func (a *Activities) StickySubmission(ctx context.Context, input models.SubmissionInput) (bool, error) {
	sub, err := a.Reddit.GetSubmission(input.SubmissionID)
	if err != nil {
		return false, err
	}
	if sub != nil && !sub.Stickied {
		if err := a.Reddit.StickySubmission(input.SubmissionID, true); err != nil {
			return false, err
		}
		activity.GetLogger(ctx).Info("Stickied", "url", "https://reddit.com"+sub.Permalink)
	}
	return true, nil
}

// UnstickySubmission unstickies a submission (idempotent).
func (a *Activities) UnstickySubmission(ctx context.Context, input models.SubmissionInput) (bool, error) {
	sub, err := a.Reddit.GetSubmission(input.SubmissionID)
	if err != nil {
		return false, err
	}
	if sub != nil && sub.Stickied {
		if err := a.Reddit.StickySubmission(input.SubmissionID, false); err != nil {
			return false, err
		}
		activity.GetLogger(ctx).Info("Unstickied", "url", "https://reddit.com"+sub.Permalink)
	}
	return true, nil
}

// LockSubmission locks a submission (idempotent).
func (a *Activities) LockSubmission(ctx context.Context, input models.SubmissionInput) (bool, error) {
	sub, err := a.Reddit.GetSubmission(input.SubmissionID)
	if err != nil {
		return false, err
	}
	if sub != nil && !sub.Locked {
		if err := a.Reddit.LockSubmission(input.SubmissionID); err != nil {
			return false, err
		}
		activity.GetLogger(ctx).Info("Locked", "url", "https://reddit.com"+sub.Permalink)
	}
	return true, nil
}

// CreateMonthlyPost creates a monthly confirmation thread (idempotent).
func (a *Activities) CreateMonthlyPost(ctx context.Context, input models.CreateMonthlyPostInput) (string, error) {
	now := time.Now().UTC()
	botName, _, err := a.Reddit.BotUser()
	if err != nil {
		return "", err
	}

	var prevTitle, prevPermalink string
	var prevCreated float64
	var prevID string
	hasPrev := false

	if input.PreviousSubmissionID != nil && *input.PreviousSubmissionID != "" {
		sub, err := a.Reddit.GetSubmission(*input.PreviousSubmissionID)
		if err != nil {
			return "", err
		}
		if sub != nil {
			prevID = sub.ID
			prevTitle = sub.Title
			prevPermalink = sub.Permalink
			prevCreated = sub.CreatedUTC
			hasPrev = true
		}
	} else {
		subs, err := a.Reddit.BotSubmissions(1)
		if err != nil {
			return "", err
		}
		if len(subs) > 0 {
			prevID = subs[0].ID
			prevTitle = subs[0].Title
			prevPermalink = subs[0].Permalink
			prevCreated = subs[0].CreatedUTC
			hasPrev = true
		}
	}

	if hasPrev {
		postDate := time.Unix(int64(prevCreated), 0).UTC()
		if postDate.Year() == now.Year() && postDate.Month() == now.Month() {
			activity.GetLogger(ctx).Info("Monthly post already exists for this month (idempotent)", "id", prevID)
			return prevID, nil
		}
	}

	if !hasPrev {
		prevTitle = "Previous monthly thread"
		prevPermalink = fmt.Sprintf("https://www.reddit.com/r/%s/", a.Cfg.SubredditName)
	}

	postTmpl, err := a.Templates.Load("monthly_post")
	if err != nil {
		return "", err
	}
	titleTmpl, err := a.Templates.Load("monthly_post_title")
	if err != nil {
		return "", err
	}

	body, err := templates.FormatString(postTmpl, map[string]any{
		"bot_name":       botName,
		"subreddit_name": a.Cfg.SubredditName,
		"previous_month_submission": map[string]any{
			"title":     prevTitle,
			"permalink": prevPermalink,
		},
		"now": now.Format(time.RFC3339),
	})
	if err != nil {
		return "", err
	}
	title := templates.FormatTitle(titleTmpl, now)

	activity.GetLogger(ctx).Info("Creating monthly post", "subreddit", a.Cfg.SubredditName)
	id, permalink, err := a.Reddit.SubmitSelfPost(title, body, a.Cfg.MonthlyPostFlairID)
	if err != nil {
		return "", err
	}
	_ = a.Reddit.SetSuggestedSort(id, "new")
	activity.GetLogger(ctx).Info("Created monthly post", "url", "https://reddit.com"+permalink)
	return id, nil
}

// SendPushoverNotification sends a Pushover notification (no-op if unconfigured).
func (a *Activities) SendPushoverNotification(ctx context.Context, message string) (bool, error) {
	appToken := os.Getenv("PUSHOVER_APP_TOKEN")
	userToken := os.Getenv("PUSHOVER_USER_TOKEN")
	if appToken == "" || userToken == "" {
		activity.GetLogger(ctx).Debug("Pushover not configured, skipping notification")
		return true, nil
	}
	form := url.Values{
		"token":   {appToken},
		"user":    {userToken},
		"message": {message},
	}
	resp, err := http.PostForm("https://api.pushover.net/1/messages.json", form)
	if err != nil {
		return false, err
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		return false, fmt.Errorf("Pushover notification failed with status %d", resp.StatusCode)
	}
	return true, nil
}

func ptrStr(p *string) string {
	if p == nil {
		return "<nil>"
	}
	return *p
}